rfcs/text/0011_global_search.md

```diff
@@ -0,0 +1,693 @@
+to server-side, as it's going the opposite way of the KP philosophy while also being an heavy breaking change.
+
+### AST parsing
+
+One option to make the `application` results 'visible' from the server-side would be to parse the client code at build time
+using AST to find all usages to `application.register` inspect the parameters, and generates a server file 
+containing the applications. The server-side `application` result provider would then just read this file and uses it 
+to return application results.
+
+However
+- At the parsing would be done at build time, we would not be able to generate entries for any 3rd party plugins
+- As entries for every existing applications would be generated, the search provider would to be able to know which
+applications are actually enabled/accessible at runtime to filter them, which is all but easy
+- It will also not contains test plugin apps, making it really hard to FTR
+- AST parsing is a complex mechanism for an already unsatisfactory alternative
+ 
+### Duplicated server-side `application.register` API 
+ 
+One other option would be to duplicate the `application.register` API on the server side, with a subset of the 
+client-side metadata.
+
+```ts
+core.application.register({
+  id: 'app_status',
+  title: 'App Status',
+  euiIconType: 'snowflake'
+});
+```
+
+This way, the applications could be searchable from the server using this server-side `applications` registry.
+
+However
+- It forces plugin developers to add this API call. In addition to be a very poor developer experience, it can also
+  very easily be forgotten, making a given app non searchable
+- client-side only plugins would need to add a server-side part to their plugin just to register their application on
+  the server side
+  
+## `GlobalSearchResult.url` could be a struct instead of a url for internal results.
+
+The initial idea for `GlobalSearchResult.url` was to have a polymorphic structure instead of just a string:
+
+```ts
+url: { absUrl: string } | { application: string; path?: string };
+```
+
+That was making it way easier to redirect to an internal result from the UI, as we could directly call
+`application.navigateTo(application, { path })`.
+
+However, that didn't bring answer for the (future) need to be able to search for and redirect to object in 
+different spaces. We could then have changed the structure to
+
+```ts
+url: { absUrl: string } | { application: string; path?: string, space?: string };
+```
+
+But this had some caveats:
+- `space` is an xpack plugin, adding this `space` property in an oss part of the code is problematic
+- The `space` API is not callable from core or oss, meaning that we would have to 'forge' the url to this space anyway
+- this is really not generic. If another plugin was to alter the basepath in another way, we would have needed to add it another property 
+
+So even if the 'parsable absolute url' approach seems fragile, it still felt better than this alternative.
```

Comment:
Would it be better to rely on a struct when available/possible and fallback to the URL approach only when a struct isn't there?

It does make it a little more complicated in that we now have two possible logic paths but I think the reduced fragility might be worth it. There's also nothing to say that just because we don't see a structured solution for the spaces problem today, doesn't mean that we won't have one someday. 

Having the struct be an option from the start I think would then encourage us to build on firmer ground in the future whenever we can.

Reply:
> it does make it a little more complicated in that we now have two possible logic paths but I think the reduced fragility might be worth it

Can you list the upsides you see with this approach?

From my point of view, we will still have to forge absolute urls from this struct when exposing the results for external consumers, therefor the 'fragile' logic will still be present. Re-introducing this polymorphic type would just introduce more complexity without removing any downsides of the other approach imho?

Also, as `GS` API will exposes a `navigateTo` method on each result on the client-side to keep this navigation logic internal, how is this impacting consumers of the `find` API?

Also, as specified, we already accept that providers return relative paths instead of absolute url, that GS will consolidate for them, meaning that

`url: /dashboard/my-dashboard` is a valid url for a `result.url`

Only upside I see is that it might be a little more clear for local result providers to return something like 

```ts
{
 app: 'dashboard',
 path: 'my-dashboard'
}
```

than `/dashboard/my-dashboard`.

In that case maybe we should differentiate the `GlobalSearchResult` type returned from the result providers from the one returned by the actual GS `find` API.

That way , providers could return urls in this `{app, path}` form that would be consolidated to absolute urls as it's already done for relative urls.

We would have

```ts
/**
 * result returned from a provider
 */
interface ProviderSearchResult {
  // abs, relative or struct representation of the result url
  url: string | { app: string; path?: string };
}

/**
 * (server-side) result returned from the search API
 */
interface GlobalSearchResult {
  // absolute url to the result
  url: string;
}

/**
 * (client-side) result returned from the search API
 */
interface GlobalSearchResult {
  // absolute url to the result
  url: string;
  // perform navigation logic to this result
  navigateTo(): void;
}
```

Reply:
> we will still have to forge absolute urls from this struct when exposing the results for external consumers, therefor the 'fragile' logic will still be present

That's true and maybe is reason enough to leave it at just always being a URL. But what I'm wondering if any reduction in using the URL is worth it? We won't be able to get rid of the absolute URL version of the results, but would reducing it's use by XX% be worth the cost of maintaining two result types?

Other than the general impression of parsing a URL for a bunch of info being fragile, I don't really have concrete examples so maybe that means the struct isn't worth introducing.

Reply:
> We won't be able to get rid of the absolute URL version of the results, but would reducing it's use by XX% be worth the cost of maintaining two result types?

I'm still unclear on what we would be gaining by avoiding string URLs. These URLs can still point to a generic endpoint that handles converting a bag of options into a redirect to an application.

For example, say the URL generator service provides a "redirect application" in the frontend that handles URLs of the form: `/app/redirect?appId=foo&options=(bar:baz,qux:t)`. A global search result provider could leverage an API provided by the URL generator to produce these URLs: `url = urlGenerator.generate({ appId: 'foo', options: { bar: 'baz', qux: true })`. This allows a result provider to have no knowledge about the final form of these URLs without needing the GS API to provide any special functionality to support this.

I'm not so worried about the parsing of URLs on the frontend since the two things we must parse are both well-known: basePath and app route. We know exactly what these two values are at all times _and_ it allows us to support spaces seamlessly.

Reply:
👍 Seems like you two have a plan here so I'm happy to go with it

Reply:
The more I think about it, the more I like the idea of having distinct types for the results returned by the providers and the results returned by the GS `find` API. 

I really wanted to allow the `{ app: string; path?: string }` format for the providers, as I think it's a more readable way to describe the path to an object, however as the applications are not known from the server-side (and as an app path is non predictable due to the possible custom `appRoute` property), we wouldn't be able to convert these to relative path for the server-side API.

So to handle all cases (abs url, path relative to the basePath, path including another basePath), I think we can have:
```ts
url: string | { path: string; prependBasePath: boolean };
```

A provider returning results from the same basePath as the request (the most common usage) would return
```ts
url: '/app/dashboard/my-dashboard'
```

A provider returning results from another instance would return
```ts
url: 'https://other-kibana/base-path/app/some-item'
```

and finally, a provider returning a result from another space or basePath would be able to return
```ts
url: { path: '/base-path/another-space/dashboard/my-dashboard', prependBasePath: false }
```

Reply:
This seems like a good compromise between dev XP usability and simplicity 👍

Reply:
Done in https://github.com/elastic/kibana/pull/64284/commits/9bf9231c090ea5ef322b7626efaa6063491b5399